gamescope: init at 3.11.33-jupiter-3.3-2

nrdxp opened the pull request:
###### Description of changes

Closes #162562

Based on @samueldr's version https://github.com/NixOS/nixpkgs/issues/162562#issuecomment-1138026849 

All I did was update to the latest tag, ~~add a pkgconfig file for stb~~, and add a small patch to stop meson from trying to load wlroots and libliftoff from it's src directory and instead rely on pkgconfig. That way we don't have to rely on copying sources as in the [original](https://github.com/Jovian-Experiments/Jovian-NixOS/blob/d27526e7917671ad3a64f823f4b95b761f96807d/pkgs/gamescope/default.nix#L72-L87).

I also figured out a way to add the `cap_sys_nice` capbility for NixOS systems to launch steam. But the change may be a bit contentious since I'd have to patch the steam derivation (since bubblewrap will fail without suid bit), so I'll save that for a seprate PR.

**Note**
~~I have not been able to successfully launch gamescope with the DRM backend from a VT. I have only been able to use it from a wayland session so far. Not sure if I just don't know the right flags to pass, or if this is a genuine upstream bug, or if it's an issue with Nvidia on a dual graphics laptop. Likely an upstream bug https://github.com/Plagman/gamescope/issues/498#issuecomment-1186386392~~ It works by passing `--disable-layers` https://github.com/Plagman/gamescope/issues/560

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
gamescope init at 3.11.33-jupiter-3.3-2

samueldr commented:
Were you able to launch it on DRM using `3.11.28-beta4`, the version in the Jovian Experiments repo?

I was able to on the Steam Deck. Untested on any other hardware. (It was how I validated that touch input was broken when running embedded, but not on DRM.)

zhaofengli commented:
Still waiting for my Steam Deck to arrive, but could the touch problem be related to the `--default-touch-mode` flag? I was looking at the stock `gamescope-session` and it seems to pass a lot of flags:

```
gamescope \
        --generate-drm-mode fixed \
        --xwayland-count 2 \
        -w 1280 -h 800 \
        --default-touch-mode 4 \
        --hide-cursor-delay 3000 \
        --max-scale 2 \
        --fade-out-duration 200 \
        -e -R "$socket" -T "$stats" \
        -O '*',eDP-1 \
        --cursor-hotspot 5,3 --cursor ...
```

nrdxp commented:
> Were you able to launch it on DRM using `3.11.28-beta4`, the version in the Jovian Experiments repo?

Unfortunately not, I've tried a number of different vesions but none launch successfully from a VT console, at least not with my Nvidia card, haven't really tested with the Intel chip.

samueldr commented on pkgs/applications/window-managers/gamescope/default.nix:
```diff
@@ -0,0 +1,82 @@
+    ];
+
+    dontUseCmakeConfigure = true;
+  }
```
`meta` is missing.

nrdxp commented:
Added the meta and removed some uneeded buildInputs. Also added a wrapper because of [this call](https://github.com/Plagman/gamescope/blob/71717b2ce38805e18f1c51d5d443873ecf41e778/src/steamcompmgr.cpp#L2373-L2374), though it's only used for a debug flag it doesn't really increase the closure size so I think it's fine.

nrdxp commented:
Finally figured out how to launch from a VT, probably useful information for others. Opened an issue to request to make the behavior automatic: Plagman/gamescope#560

SuperSandro2000 commented on pkgs/applications/window-managers/gamescope/default.nix:
```diff
@@ -0,0 +1,95 @@
+
+    patches = [./use-pkgconfig.patch];
+
+    buildInputs = [
+      xorg.libXdamage
+      xorg.libXcomposite
+      xorg.libXrender
+      xorg.libXext
+      xorg.libXxf86vm
+      xorg.libXtst
+      xorg.libXres
+      xorg.libXi
+      libdrm
+      libliftoff
+      vulkan-loader
+      glslang
+      SDL2
+      wayland
+      wayland-protocols
+      wlroots
+      xwayland
+      libseat
+      libinput
+      libxkbcommon
+      udev
+      pixman
+      pipewire
+      libcap
+      stb
+    ];
+
+    nativeBuildInputs = [
+      cmake
+      meson
+      pkg-config
+      ninja
+      makeBinaryWrapper
+    ];
```
```suggestion
    nativeBuildInputs = [
      cmake
      meson
      pkg-config
      ninja
      makeBinaryWrapper
    ];

    buildInputs = [
      xorg.libXdamage
      xorg.libXcomposite
      xorg.libXrender
      xorg.libXext
      xorg.libXxf86vm
      xorg.libXtst
      xorg.libXres
      xorg.libXi
      libdrm
      libliftoff
      vulkan-loader
      glslang
      SDL2
      wayland
      wayland-protocols
      wlroots
      xwayland
      libseat
      libinput
      libxkbcommon
      udev
      pixman
      pipewire
      libcap
      stb
    ];
```

SuperSandro2000 commented on pkgs/applications/window-managers/gamescope/default.nix:
```diff
@@ -0,0 +1,95 @@
+      hash = "sha256-KAvOmxYSwLA6/zOppDzgI7z8+XCMMEPol1vo/SDuRXE=";
+    };
+
+    patches = [./use-pkgconfig.patch];
```
```suggestion
    patches = [ ./use-pkgconfig.patch ];
```

SuperSandro2000 commented on pkgs/applications/window-managers/gamescope/default.nix:
```diff
@@ -0,0 +1,95 @@
+    ];
+
+    # --debug-layers flag expects these in the path
+    makeWrapperArgs = ["--prefix" "PATH" ":" "${with xorg; lib.makeBinPath [xprop xwininfo]}"];
```
```suggestion
    makeWrapperArgs = ["--prefix" "PATH" ":" "${with xorg; lib.makeBinPath [ xprop xwininfo ]}"];
```

SuperSandro2000 commented on pkgs/applications/window-managers/gamescope/default.nix:
```diff
@@ -0,0 +1,95 @@
+{
+  stdenv,
+  fetchFromGitHub,
+  meson,
+  pkg-config,
+  cmake,
+  ninja,
+  xorg,
+  libdrm,
+  vulkan-loader,
+  wayland,
+  wayland-protocols,
+  libxkbcommon,
+  libcap,
+  SDL2,
+  pipewire,
+  udev,
+  pixman,
+  libinput,
+  libseat,
+  xwayland,
+  glslang,
+  stb,
+  wlroots,
+  libliftoff,
+  lib,
+  makeBinaryWrapper,
+}: let
+  pname = "gamescope";
+  version = "3.11.33-beta1";
+in
+  stdenv.mkDerivation {
```
```suggestion
{ stdenv
, fetchFromGitHub
, meson
, pkg-config
, cmake
, ninja
, xorg
, libdrm
, vulkan-loader
, wayland
, wayland-protocols
, libxkbcommon
, libcap
, SDL2
, pipewire
, udev
, pixman
, libinput
, libseat
, xwayland
, glslang
, stb
, wlroots
, libliftoff
, lib
, makeBinaryWrapper
}:

let
  pname = "gamescope";
  version = "3.11.33-beta1";
in
stdenv.mkDerivation {
```

SuperSandro2000 reviewed (commented):
are you using alejandra to format? If so I highly recommend to switch back to nixpkgs-fmt because it creates code that is not compatible with the style in nixpkgs.

nrdxp commented:
> are you using alejandra to format? If so I highly recommend to switch back to nixpkgs-fmt because it creates code that is not compatible with the style in nixpkgs.

ah yeah, didn't even think about that, but your right, atm the format of nixpkgs is more in line with nixpkgs-fmt output. I'll reformat in a bit

bjornfor commented on pkgs/applications/window-managers/gamescope/default.nix:
```diff
@@ -0,0 +1,95 @@
+  ];
+
+  nativeBuildInputs = [
+    cmake
```
It needs both cmake and meson?

SuperSandro2000 replied:
and ninja? also nativeBuildInputs are usually sorted before buildInputs

nrdxp replied:
Hmm I don't think `cmake` is required actually, must have just have been left over from the original I copied. I made a pass to remove unused `buildInputs` but I neglected the `nativeBuildInputs`.  But yes both meson and ninja are legitimately needed.

SuperSandro2000 commented on pkgs/applications/window-managers/gamescope/default.nix:
```diff
@@ -0,0 +1,95 @@
+  version = "3.11.33-beta1";
+in
+stdenv.mkDerivation {
+  inherit pname version;
+  src = fetchFromGitHub {
```
```suggestion
  inherit pname version;

  src = fetchFromGitHub {
```

SuperSandro2000 commented on pkgs/applications/window-managers/gamescope/default.nix:
```diff
@@ -0,0 +1,95 @@
+    makeBinaryWrapper
+  ];
+
+  # --debug-layers flag expects these in the path
+  makeWrapperArgs = [ "--prefix" "PATH" ":" "${with xorg; lib.makeBinPath [xprop xwininfo]}" ];
+
+  postInstall = ''
+    wrapProgram "$out/bin/gamescope" ''${makeWrapperArgs[@]}
+  '';
```
```suggestion
  postInstall = ''
    # --debug-layers flag expects these in the path
    wrapProgram "$out/bin/gamescope" \
      --prefix PATH : ${with xorg; lib.makeBinPath [ xprop xwininfo ] }
  '';
```

nrdxp commented:
any reviewers so far care to approve?

rapenne-s reviewed (approved):
looks fine to me

tried with nvidia and intel, on Wayland and X
